File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    private static long largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        long maxArea = 0L;

        for (int i=0; i<=heights.length; i++) {
            int currentHeight = (i == heights.length) ? 0 : heights[i];

            while (!stack.isEmpty() && heights[stack.peek()] > currentHeight) {
                int height = heights[stack.pop()];
                int weight = stack.isEmpty() ? i : i - stack.peek() - 1;

                maxArea = Math.max(maxArea, (long) weight * height);
            }

            stack.push(i);
        }

        return maxArea;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        while (true) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            if (n == 0) break;

            int[] heights = new int[n];
            for (int i=0; i<n; i++) {
                heights[i] = Integer.parseInt(st.nextToken());
            }

            System.out.println(largestRectangleArea(heights));
        }
    }
}

File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.kt
import java.util.Stack

/**
 * p6549 히스토그램에서 가장 큰 직사각형
 */
fun largestRectangleArea(heights: List<Int>): Long {
    val stack = Stack<Int>()
    var maxArea = 0L

    for (i in 0 .. heights.size) {
        val currentHeight = if (i == heights.size) 0 else heights[i]

        while (stack.isNotEmpty() && heights[stack.peek()] > currentHeight) {
            val height = heights[stack.pop()]
            val width = if (stack.isEmpty()) i else i - stack.peek() - 1
            maxArea = maxOf(maxArea, width.toLong() * height.toLong())
        }

        stack.push(i)
    }

    return maxArea
}

//fun largestSquareArea(heights: List<Int>): Int {
//    val stack = Stack<Int>()
//    var maxArea = 0
//
//    for (i in 0 .. heights.size) {
//        val currentHeight = if (i == heights.size) 0 else heights[i]
//
//        while (stack.isNotEmpty() && heights[stack.peek()] > currentHeight) {
//            val height = heights[stack.pop()]
//            val width = if (stack.isEmpty()) i else i - stack.peek() - 1
//
////             정사각형 조건 : 너비, 높이 중 작은 값을 한 변으로
//            val side = minOf(width, height)
//            maxArea = maxOf(maxArea, side * side)
//        }
//
//        stack.push(i)
//    }
//
//    return maxArea
//}

fun main() {
    while (true) {
        val input = readln().split(" ").map { it.toInt() }
        if (input[0] == 0) break
        val heights = input.subList(1, input.size)
        println(largestRectangleArea(heights))
//        println(largestSquareArea(heights))
    }
}